### PR TITLE
Named and destructable `self`

### PR DESCRIPTION
**Pull request by hauleth**
[Rendered](https://github.com/hauleth/rfcs/blob/self-param-as-sugar/text/0000-named-and-destructurable-self.md).

### CODE REVIEW

**Comment by chris-morgan** [comment]
Destruction and destructuring are two rather different things.

File: text/0000-named-and-destructable-self.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+- Feature Name: named_and_destructable_self
+- Start Date: 2015-03-02
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Allow destructing assignment for `self` argument.
```
Comment (jaredly) [inline text/0000-named-and-destructable-self.md]: destructuring
and all the other places

**Comment by kennytm** [comment]
The "Rendered" is a dead link.

**Comment by pnkfelix** [comment]
(fixed link)

**Comment by pnkfelix** [comment]
Thank you for posting this.

I actually believe the intention, during the UFCS design, was to add this feature (or something very much like it).  In particular, you can today already write:

```rust
struct Foo(u32, String);

impl Foo {
    fn bar(x: Self) -> String {
        x.1
    }
}
```

(note however, that one is currently forced to call the above via the syntax `Foo::bar(..)`; the expression `f1.bar()` for `f1: Foo` does not yet work; see the comment thread attached to #395 for a little more discussion on this point.)

----

Anyway, we do already have a start to what you describe.  Of course, in addition to the method call issue noted above, it also does not yet allow destructuring. (But I suspect this was an oversight, not a concious design choice.)

@nikomatsakis can you comment on whether the [UFCS RFC](https://github.com/rust-lang/rfcs/blob/master/text/0132-ufcs.md) needs an amendment?

**Comment by pnkfelix** [comment]
reassigning to @nikomatsakis

**Comment by nikomatsakis** [comment]
Indeed, the already accepted [RFC 48](https://github.com/rust-lang/rfcs/blob/master/text/0048-traits.md) permitted this sort of syntax. That was not implemented. I have some concerns that we may not be able to implement it backwards compatibly -- in particular, we currently use the `self` keyword to designate methods that should be considered for method dispatch, and growing that set can create ambiguities. Now that we are moving towards 1.0 we'll have to see what is possible and desirable in the future. Nonetheless, I'm going to close this RFC since it is already the official plan, and the limiting factor is more backwards compatibility and implementation time and effort. Thanks very much for the suggestion!

**Pull request closed without merging**